File: Core/Swift/queue.swift
class QueueNode<T> {
  public var value: T;
  public var next: QueueNode?;
  init(value: T) {
    self.value = value;
  }
}

class Queue<T> {
  public var first: QueueNode<T>?;

  // Remove first in line.
  public func pop() -> T? {
    if (self.first != nil) {
      let tmp = self.first;
      self.first = self.first?.next;
      return tmp!.value;
    }
    return nil;
  }

  public func peek() -> QueueNode<T>? {
    return self.first
  }

  public func add(value: T) -> T? {
    let newNode = QueueNode<T>(value: value);
    if (self.first == nil) {
      self.first = newNode;
      return newNode.value;
    }
    guard var node = self.first else {
      return nil;
    }
    while let next = node.next {
      node = next
      if (next == nil) {
        node.next = newNode;
        return node.next!.value;
      }
    }
    return nil;
  }

  public func isEmpty() -> Bool {
    return self.first == nil;
  }
}

File: Core/Python/queue.py
# Queue implementation.

class Queue:
  first = None
  last = None

  def add(self, value):
    # add using last node
    new_node = QueueNode(value)
    if (self.first == None):
      self.first = new_node
    if (self.last == None):
      self.last = new_node
    self.last.next = new_node
    new_node.prev = self.last
    self.last = new_node

  def remove(self):
    # Remove from start
    if (self.first):
      removed_node = self.first
      self.first = self.first.next
      return removed_node
    raise Exception('Queue has no first to remove.')

  def peek(self):
    return self.first

  def is_empty(self):
    return (self.first == None)
      

class QueueNode:
  prev = None
  next = None
  def __init__(self, value):
    self.value = value